### PR TITLE
roachpb: modernize AmbiguousResultError

### PR DESCRIPTION
**Pull request by tbg**
AmbiguousResultError previously had a `WrappedErr *Error` field.
Wrapping `Error` inside of itself is a really bad idea! Luckily,
it wasn't used by anyone, but it is being used now, with replica
circuit breakers.

Before we cement this in a production release, give this error type a
do-over and let it "properly" wrap an error. I performed an audit to
make sure we were always checking for ambiguous results first, i.e.
looked for bugs of the form

```go
switch err.(type) {
  case *roachpb.SomeRetriableError:
    // Retry request.
    continue
  case *AmbiguousResultError:
    return err
}
```

where `err` might be an `AmbiguousResultError` wrapping a
`NotRetriableError`. I didn't find anything; we always carry out the
ambiguous result check first. Also, very few errors are wrapped in an
ambiguous result, and most of them could actually take precedence. If
for example an ambiguous result wraps a lease error, the result is not
ambiguous. We don't try to exploit that at all in this commit, and
likely never will since this isn't a common situation.

Release justification: improvement to new functionality (#33007)
Release note: None

### CODE REVIEW

**Comment by tbg** [comment]
PS: search the diff for "WrappedErr" to see where we were using it.

File: pkg/roachpb/errors.go [hunk 1]
```diff
@@ -654,27 +654,65 @@ func (e *RangeKeyMismatchError) AppendRangeInfo(ctx context.Context, ris ...Rang
...
+//   err := NewAmbiguousResultErrorf("foo")
+//   err.EncodedError = errors.EncodeError(context.Background(), &FooError{})
+//   if errors.HasType(err, (*FooError)(nil)) {
+//     /* handle FooError */
+//   } else if errors.HasType(err, (*AmbiguousResultError)(nil)) {
+//     /* handle ambiguous result */
+//   }
+//
+// In other words, we're making unwrapping through an AmbiguousResultError an
+// explicit choice.
```
Comment (erikgrinaker) [inline pkg/roachpb/errors.go:704]: It isn't clear to me why this is always the right choice. Shouldn't it be up to the caller to decide on precedence, as different callers may care about different things?

Apart from that, this all seems fine.

Comment (tbg) [reply pkg/roachpb/errors.go:704]: I agree, just worried we're not being consistent. But let me do another pass, should be easy to audit.

Comment (tbg) [reply pkg/roachpb/errors.go]: Nothing really turned up, thanks for the nudge. Updated & rebased on top of #77365 (which is currently in CI), PTAL.

**Review by erikgrinaker — APPROVED** [review]
LGTM, thanks!

**Comment by tbg** [comment]
bors r=erikgrinaker
Flakes are not mine.

**Pull request merged**